---
title: "Quality Control for scRNA and scATAC seq data"
subtitle: "Merge the data"
author: "Ximing Ran"
date: "2025-05-01"
output:
  pdf_document:
    toc: true
    toc_depth: 3
    number_sections: false
  html_document:
    # code_folding: hide
    toc: true
    toc_float: true
hitheme: tomorrow
highlighter: highlight.js

---

```{r setup, include=FALSE}
# load libraries
library(tidyverse)
library(knitr)
set.seed(2025)

knitr::opts_chunk$set(
  # echo = FALSE,
  # warning = FALSE,
  # message = FALSE,
  fig.width = 9,
  fig.height = 4,
  fig.path = "./results/Merge-analysis-plot/"
)

theme_set(theme_bw() + theme(legend.position = "bottom"))
knitr::kable(head(mtcars[, 1:4]), "simple")
```

```{r}
library(Seurat)
library(tidyverse)

```



```{r local_function_load}
# load function from local files

```


```{r local_reference_load}
# load data from local files
library(Signac)
library(Seurat)

```


\newpage

# 1. Read the QC Data


```{r load_data_TSC}
options(Seurat.object.assay.version = "v5")

# load data
sample_seurat_list <- readRDS(
  here::here("data", "processed-data",
             "sample_seurat_list_2025-05-01.rds")
)

sample_meta <- readRDS(
  here::here("data", "processed-data",
             "sample_meta_2025-05-01.rds")
)



for (samp in names(sample_seurat_list)) {
  cat("Sample:", samp, "\n")
  print(head(colnames(sample_seurat_list[[samp]]), 10))
  cat("\n")
}


```

\newpage
# 2. Merge the data


```{r merge_samples, fig.width=12, fig.height=8}
library(Seurat)
library(harmony)
library(Signac)
library(GenomicRanges)
library(dplyr)

names(sample_seurat_list) <- sample_meta

# 1) Pre-process each sample’s RNA (normalize → HVG → scale → PCA)
sample_seurat_list <- lapply(sample_seurat_list, function(obj) {
  DefaultAssay(obj) <- "RNA"
  obj %>%
    NormalizeData(assay = "RNA", layer = "counts.Gene Expression") %>%
    FindVariableFeatures(assay = "RNA", layer = "counts.Gene Expression", 
                         nfeatures = 3000) %>%
    ScaleData(assay = "RNA", layer = "counts.Gene Expression") %>%
    RunPCA(assay = "RNA", reduction.name = "pca", npcs = 50)
})



# 2) Merge all samples into one Seurat object (RNA + ATAC assays carried along)
merged <- merge(
  x            = sample_seurat_list[[1]],
  y            = sample_seurat_list[-1],
  add.cell.ids = names(sample_seurat_list) ,
  project      = "MergedProject"
)

# 3) Re-run PCA on the merged RNA (just in case)
DefaultAssay(merged) <- "RNA"
merged <- merged  %>%
  NormalizeData(assay = "RNA", layer = "counts.Gene Expression") %>%
  FindVariableFeatures(assay = "RNA", layer = "counts.Gene Expression", 
                       nfeatures = 3000) %>%
  ScaleData(assay = "RNA", layer = "counts.Gene Expression") %>%
  RunPCA(assay = "RNA", reduction.name = "pca", npcs = 50)

# 4) Harmony integration on the RNA PCs, correcting for your batch/group
merged <- RunHarmony(
  object        = merged,
  group.by.vars = "orig.ident",    
  reduction.use = "pca",
  dims.use      = 1:50
)

# 5) UMAP & clustering on the harmony embedding
merged <- merged %>%
  RunUMAP(reduction = "harmony", dims = 1:50) %>%
  FindNeighbors(reduction = "harmony", dims = 1:50) %>%
  FindClusters(resolution = 0.5)

# visualize RNA integration
p<-DimPlot(merged, reduction = "umap", group.by = "orig.ident") + 
  ggtitle("RNA Harmony UMAP")
ggsave(p, filename = "./results/Merge-analysis-plot/RNA_Harmony_UMAP.png",
       width = 8, height = 6, dpi = 300)

p<-DimPlot(merged, reduction = "umap", group.by = "orig.ident", 
           split.by = "orig.ident") + ggtitle("RNA Harmony UMAP")
ggsave(p, filename = "./results/Merge-analysis-plot/RNA_Harmony_UMAP_split.png",
       width = 20, height = 5, dpi = 300)

# show the cluster results
p<-DimPlot(merged, reduction = "umap", group.by = "seurat_clusters") + 
  ggtitle("RNA Harmony UMAP Clusters")
ggsave(p, filename = "./results/Merge-analysis-plot/RNA_Harmony_UMAP_clusters.png",
       width = 8, height = 6, dpi = 300)


# 

# 6) Collapse all ATAC peak ranges at once with reduce()
all_ranges <- lapply(sample_seurat_list, function(obj) obj@assays$ATAC@ranges)
peaks      <- reduce(unlist(as(all_ranges, "GRangesList"))) 
peaks      <- peaks[width(peaks) > 20 & width(peaks) < 10000]

# peaks is now a single GRanges of your union peaks
# you can rebuild the ATAC assay if you like:
DefaultAssay(merged) <- "ATAC"
frag    <- Fragments(merged)
counts  <- FeatureMatrix(fragments = frag, 
                         features = peaks, 
                         cells = colnames(merged))
merged[["ATAC"]] <- CreateChromatinAssay(counts = counts, 
                                         ranges = peaks, 
                                         fragments = frag, 
                                         genome = "hg38")

DefaultAssay(merged) <- "RNA"

# save the data
saveRDS(merged, here::here("data", "processed-data","merged_2025-05-01.rds"),
        compress = FALSE)
```



\newpage

